### PR TITLE
memoize compute extents

### PR DESCRIPTION
**Pull request by hellochar**
Memoize calls to computing the plot extents (which does a linear pass through the data), improving perf. This also lets us delete a bunch of _update methods and move away from the "update" paradigm

### CODE REVIEW

File: src/memoize/memoize.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
+/**
+ * Copyright 2014-present Palantir Technologies
```
Comment (themadcreator) [inline src/memoize/memoize.ts:2]: technically i think these should be 2017-present

**Review by themadcreator — APPROVED** [review]

File: src/memoize/memThunk.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    arg2: Thunk<I2>,
+    arg3: Thunk<I3>,
+    compute: (this: void, arg1: I1, arg2: I2, arg3: I3) => O,
+): MemoizedFunction<Thunk<O>>;
+export function memThunk<I1, I2, I3, I4, O>(
+    arg1: Thunk<I1>,
+    arg2: Thunk<I2>,
+    arg3: Thunk<I3>,
+    arg4: Thunk<I4>,
+    compute: (this: void, arg1: I1, arg2: I2, arg3: I3, arg4: I4) => O,
```
Comment (themadcreator) [inline src/memoize/memThunk.ts:31]: you know what we need? variadic template types. dang.

File: src/memoize/signature.ts [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+export class ReferenceSignature extends Signature {
+    constructor(private ref: any) {
+        super();
+    }
+
+    isSignatureDifferent(other: ReferenceSignature) {
+        return this.ref !== other.ref;
+    }
+}
+
```
Comment (themadcreator) [inline src/memoize/signature.ts:145]: `export type ISignatureRecord = Record<string, Signature>;`

File: src/memoize/signature.ts [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+/**
+ * Base signature. Subclasses should implement isSignatureDifferent. All classes
+ * should be immutable.
+ *
+ * Users should only call `isDifferent`, not `isSignatureDifferent`.
+ */
+export abstract class Signature {
+    public isDifferent(other: Signature): boolean {
+        // debugger;
```
Comment (themadcreator) [inline src/memoize/signature.ts]: stray debugger

File: src/memoize/signature.ts [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+ * Generic signature factory - pass any value and get a signature for it.
+ *
+ * Datasets and Scales are handled specially - see their respective signing methods.
+ *
+ * If the input is already a signature, simply return it.
+ *
+ * @param a
+ * @returns {Signature}
+ */
+export function sign<T>(a: T): Signature {
```
Comment (themadcreator) [inline src/memoize/signature.ts]: since you're not using T anywhere else, isn't this as good as `any` ?

File: src/memoize/memoize.ts [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+    /**
+     * Turn on perf logging. Useful for debugging.
+     */
+    logPerformance(): this;
+}
+
+export type MemoizedFunction<F extends Function> = F & IMemoizedFunctionMethods;
+
+export function memoize<F extends Function>(compute: F): MemoizedFunction<F> {
```
Comment (themadcreator) [inline src/memoize/memoize.ts]: alright, so this is the actual memoize function that will be the main first thing people read. i'd like the comment to tell me:

how to use it (short code example).
what sort of pitfalls there are (vis a vis signatures). a warning that arguments that are not shallow comparable may need a custom signature.
also,  what memoization locking is and why you would want to use it.

File: quicktests/overlaying/tests/realistic/scatter_bar.js [hunk 7]
```diff
@@ -25,7 +25,7 @@ function run(svg, data, Plottable) {
   var bars = new Plottable.Plots.Rectangle()
     .addDataset(ds1)
     .x(function (d) { return d.overall; }, xScale)
-    .x2(function(d) { return d.leadership; }, xScale)
+    .x2(function(d) { return d.leadership; })
```
Comment (themadcreator) [inline quicktests/overlaying/tests/realistic/scatter_bar.js:28]: why was this necessary? doesn't calling `.x2` here re-use the `xScale` from the call to `.x` above anyway?

File: src/memoize/memoize.ts [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+        if (locked) {
+            throw new Error("Locking an already locked memoize function!");
+        }
+        locked = true;
+        const retVal = cb.apply(this);
+        locked = false;
+        return retVal;
+    };
+
+    memoizeFn.logPerformance = function() {
```
Comment (themadcreator) [inline src/memoize/memoize.ts]: since this is debugging only, i doubt you'd care to turn it off, but it might be worth taking an optional argument.

Comment (themadcreator) [reply src/memoize/memoize.ts]: `.logPerformance(logit = true)`

File: src/plots/plot.ts [hunk 9]
```diff
@@ -437,56 +434,49 @@ export class Plot extends Component {
...
-
-    if (scale == null) {
-      return [];
-    }
-
-    let data = dataset.data();
-    if (filter != null) {
-      data = data.filter((d, i) => filter(d, i, dataset));
+  protected getExtentsForAttr(attr: string) {
+    if (this._attrExtents[attr] == null) {
```
Comment (themadcreator) [inline src/plots/plot.ts:465]: does this ever need to be overrided or deleted once it's defined?

File: src/plots/plot.ts [hunk 10]
```diff
@@ -837,7 +827,7 @@ export class Plot extends Component {
     return {};
   }
 
-  protected static _scaledAccessor<D, R>(binding: Plots.IAccessorScaleBinding<D, R>) {
+  public static _scaledAccessor<D, R>(binding: Plots.IAccessorScaleBinding<D, R>) {
```
Comment (themadcreator) [inline src/plots/plot.ts]: public _methodName ? we should rename without underscore. or can we get away with protected?

File: src/plots/plot.ts [hunk 11]
```diff
@@ -859,3 +849,24 @@ export class Plot extends Component {
     return this._animate && this._dataChanged;
   }
 }
+
+function _computeExtent(
```
Comment (themadcreator) [inline src/plots/plot.ts]: no need to underscore for unexported functions, i think

Comment (themadcreator) [reply src/plots/plot.ts]: or do we do that elsewhere in plottable?

File: src/plots/stackedAreaPlot.ts [hunk 12]
```diff
@@ -19,8 +20,23 @@ import { Plot } from "./plot";
...
 export class StackedArea<X> extends Area<X> {
   private _stackingOrder: Utils.Stacking.IStackingOrder;
-  private _stackingResult: Utils.Stacking.StackingResult;
-  private _stackedExtent: number[];
+  private _stackingResult: Thunk<Utils.Stacking.StackingResult> = memThunk(
+      () => this.datasets(),
+      () => this.x().accessor,
+      () => this.y().accessor,
+      () => this._stackingOrder,
+      (datasets, keyAccessor, valueAccessor, stackingOrder) => {
```
Comment (themadcreator) [inline src/plots/stackedAreaPlot.ts:28]: this is basically reselect's API. should we just us reselect?

File: src/plots/stackedBarPlot.ts [hunk 13]
```diff
@@ -7,12 +7,11 @@ import * as Typesettable from "typesettable";
 
 import { Dataset } from "../core/dataset";
 import { Formatter, identity } from "../core/formatters";
-import { Bounds, IAccessor, Point, SimpleSelection } from "../core/interfaces";
-import { Scale } from "../scales/scale";
+import { Bounds, Point, SimpleSelection } from "../core/interfaces";
+import { memThunk, Thunk } from "../memoize/index";
```
Comment (themadcreator) [inline src/plots/stackedBarPlot.ts]: don't think you need the `/index` ?

File: test/memoize/memoizeTests.ts [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+import * as sinon from "sinon";
+
+import * as Plottable from "../../src";
+
+import { Dataset } from "../../src/core/dataset";
+import { memoize } from "../../src/memoize/memoize";
+import { IAccessorScaleBinding } from "../../src/plots/commons";
+import { Plot } from "../../src/plots/plot";
+
+describe("memoize()", () => {
```
Comment (themadcreator) [inline test/memoize/memoizeTests.ts:11]: 💯

File: src/plots/stackedAreaPlot.ts (hunk 12 above)
Comment (themadcreator) [reply src/plots/stackedAreaPlot.ts:28]: i guess it won't quite work if we don't do immutable objects

**Review by themadcreator — CHANGES_REQUESTED** [review]
some minor requests and questions

File: quicktests/overlaying/tests/realistic/scatter_bar.js (hunk 7 above)
Comment (hellochar) [reply quicktests/overlaying/tests/realistic/scatter_bar.js:28]: The second argument to `x2` is now the `postScale` function; this was breaking the examples

File: src/plots/plot.ts (hunk 9 above)
Comment (hellochar) [reply src/plots/plot.ts:465]: Nope, this should be defined the same way for all attrs (and will always be defined that way)

File: src/plots/stackedAreaPlot.ts (hunk 12 above)
Comment (hellochar) [reply src/plots/stackedAreaPlot.ts:28]: Yeah we basically rewrote reselect

File: quicktests/overlaying/tests/realistic/scatter_bar.js (hunk 7 above)
Comment (themadcreator) [reply quicktests/overlaying/tests/realistic/scatter_bar.js:28]: oh, okay

File: src/plots/stackedAreaPlot.ts (hunk 12 above)
Comment (themadcreator) [reply src/plots/stackedAreaPlot.ts:28]: rereselect

**Review by themadcreator — APPROVED** [review]

**Pull request merged**